### PR TITLE
[WIP] Tree Transaction Backfiller

### PR DESCRIPTION
**Pull request by kespinola**
resolves https://github.com/metaplex-foundation/digital-asset-rpc-infrastructure/issues/150

## Tasks
- [X]  Crawl all trees and transactions
- [x] Write `tree_transaction` records to db
- [X] Pickup on last processed transaction
- [x] Forward transactions to TXNFILL messenger stream
- [x] Logging
- [ ] Command progress and reports
- [ ] Performance and reliability metrics
- [ ] Ops manifests
- [ ] Documentation

### CODE REVIEW

File: tree_backfiller/src/tree.rs [hunk 1]
```diff
@@ -0,0 +1,202 @@
...
+
+    Ok(client
+        .get_program_accounts_with_config(&id(), config)
+        .await?
+        .into_iter()
+        .filter_map(|(pubkey, account)| TreeResponse::from_rpc(pubkey, account).ok())
+        .collect())
+}
+
+pub async fn crawl(
```
Comment (danenbm) [inline tree_backfiller/src/tree.rs:120]: I know this is all still WIP and I haven't fully studied it all, but at first glance seems like `crawl` could be a method of `TreeResponse`?

File: nft_ingester/src/stream.rs
[the remark this answers is not in the document]
Comment (kespinola) [reply nft_ingester/src/stream.rs:1]: Definitely, I'll be removing any erroneous changes like formatting before merging.

File: tree_backfiller/src/tree.rs (hunk 1 above)
Comment (kespinola) [reply tree_backfiller/src/tree.rs:120]: I can. I had that thought yesterday as well.

File: nft_ingester/src/program_transformers/bubblegum/db.rs [hunk 2]
```diff
@@ -71,9 +133,13 @@ where
...
-        let mut audit_item: cl_audits::ActiveModel = item.clone().into();
-        audit_item.tx = Set(txn_id.to_string());
-        audit_item.instruction = Set(Some(instruction.to_string()));
+        let audit_item: Option<cl_audits::ActiveModel> = if cl_audits {
+            let mut ai: cl_audits::ActiveModel = item.clone().into();
+            ai.tx = Set(txn_id.to_string());
+            Some(ai)
+        } else {
+            None
+        };
```
[the remark this answers is not in the document]
Comment (kespinola) [reply nft_ingester/src/program_transformers/bubblegum/db.rs:142]: Ok. This is based off a staging branch of Tritons. Will get git in a good spot.

File: nft_ingester/src/program_transformers/bubblegum/mod.rs [hunk 3]
```diff
@@ -96,6 +96,18 @@ where
         }
         _ => debug!("Bubblegum: Not Implemented Instruction"),
     }
+
+    // TODO: assuming tree update available on all transactions but need to confirm.
+    if let Some(tree_update) = &parsing_result.tree_update {
```
[the remark this answers is not in the document]
Comment (kespinola) [reply nft_ingester/src/program_transformers/bubblegum/mod.rs:103]: Ok what information changes in the index in this case? Its an instruction only change correct? Nothing in account updates causes the same phase shift.

[the remark this answers is not in the document]
Comment (danenbm) [reply nft_ingester/src/program_transformers/bubblegum/mod.rs:103]: Yeah maybe decompress doesn't truly need to be backfilled, or perhaps even indexed at all.  It looks like everything set in `decompress` indexing is also set by token-metadata account-based indexing.

**Pull request merged**